drivers/serial/uart_nrfx_uarte.c

```diff
@@ -1040,69 +1120,56 @@ static int uarte_nrfx_poll_in(const struct device *dev, unsigned char *c)
  */
 static void uarte_nrfx_poll_out(const struct device *dev, unsigned char c)
 {
-	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
 	struct uarte_nrfx_data *data = get_dev_data(dev);
-	atomic_t *lock;
-
-#ifdef CONFIG_UART_ASYNC_API
-
-	if (data->async) {
-		while (data->async->tx_buf) {
-			/* If there is ongoing transmission, and we are in
-			 * isr, then call uarte interrupt routine, otherwise
-			 * busy wait until transmission is finished.
-			 */
-			if (k_is_in_isr()) {
-				uarte_nrfx_isr_async(dev);
-			}
-		}
-		/* Use tx_size as lock, this way uarte_nrfx_tx will
-		 * return -EBUSY during poll_out.
-		 */
-		lock = &data->async->tx_size;
-	} else
+	bool isr_mode = k_is_in_isr() || k_is_pre_kernel();
+	NRF_UARTE_Type *uarte = get_uarte_instance(dev);
+	int key;
+
+	if (isr_mode) {
+		while (1) {
+			key = irq_lock();
+			if (is_tx_ready(dev)) {
+#if CONFIG_UART_ASYNC_API
+				if (data->async->tx_size &&
+					data->async->tx_amount < 0) {
+					data->async->tx_amount =
+						nrf_uarte_tx_amount_get(uarte);
+				}
 #endif
-		lock = &data->poll_out_lock;
```

Comment:
`poll_out_lock` should be also removed from the `uarte_nrfx_data` structure.

Reply:
done